File: compiler/fuzz/fuzz_parser.ml
(* SPDX-License-Identifier: PMPL-1.0-or-later *)
(* Fuzz target for the Tangle (OCaml) parser.
 *
 * Invariant: the parser must NEVER crash on ANY input. It should raise
 * Parser.Error or Lexer.Lexer_error, never an uncaught exception.
 *
 * Tangle uses ocamllex (lib/lexer.mll) + Menhir (lib/parser.mly).
 * This harness feeds random strings through the full lex+parse pipeline.
 *
 * Strategy: 50% raw random bytes, 50% structured inputs mixing
 * TANGLE keywords (def, weave, strands, match, etc.) to achieve
 * deeper parser coverage.
 *
 * Run with:
 *   dune exec fuzz/fuzz_parser.exe
 *)

let fuzz_one_input (input : string) : unit =
  let lexbuf = Lexing.from_string input in
  (try
     let _ = Parser.program Lexer.token lexbuf in
     ()
   with
   | Parser.Error -> ()
   | Lexer.Lexer_error _ -> ()
   | _ -> ())

(* Simple PRNG-based random string generator. *)
let random_bytes (rng : Random.State.t) (max_len : int) : string =
  let len = Random.State.int rng (max_len + 1) in
  let buf = Bytes.create len in
  for i = 0 to len - 1 do
    Bytes.set buf i (Char.chr (Random.State.int rng 256))
  done;
  Bytes.to_string buf

(* Fragments biased toward TANGLE syntax for deeper parser coverage.
 * TANGLE is a braid/knot theory language with vertical composition (.),
 * horizontal tensor (|), pipeline (>>), and braid generators. *)
let interesting_fragments =
  [| (* Keywords *)
     "def"; "weave"; "into"; "yield"; "strands"; "compute"; "assert";
     "match"; "with"; "end"; "let"; "in";
     (* Braid operations *)
     "close"; "mirror"; "reverse"; "simplify"; "cap"; "cup"; "twist";
     (* Operators *)
     "."; "|"; ">>"; "=="; "~"; "+"; "-"; "*"; "/";
     "->"; "=>";
     (* Delimiters *)
     "("; ")"; "["; "]"; "{"; "}"; ","; ";"; ":";
     (* Braid generators *)
     "s1"; "s2"; "s3"; "s1^-1"; "s2^-1"; "s3^-1";
     (* Literals *)
     "42"; "0"; "3.14"; "1e5"; "true"; "false";
     "\"hello\"";
     (* Identifiers *)
     "x"; "y"; "braid"; "knot"; "link"; "f"; "g"; "_";
     (* Whitespace *)
     " "; "\t"; "\n"; "\r";
     (* Comments *)
     "(* comment *)"; "// line\n";
     (* Structured patterns *)
     "def f x = x . x";
     "let b = s1 | s2 in b >> b";
     "match x with | a -> a end";
     "weave [s1, s2, s3] into knot";
  |]

let random_input (rng : Random.State.t) (max_len : int) : string =
  if Random.State.bool rng then
    random_bytes rng max_len
  else begin
    let buf = Buffer.create max_len in
    let target_len = Random.State.int rng (max_len + 1) in
    while Buffer.length buf < target_len do
      let frag = interesting_fragments.(
        Random.State.int rng (Array.length interesting_fragments)
      ) in
      Buffer.add_string buf frag
    done;
    Buffer.contents buf
  end

let () =
  let rng = Random.State.make_self_init () in
  let iterations =
    try int_of_string (Sys.getenv "FUZZ_ITERATIONS")
    with _ -> 100_000
  in
  Printf.printf "Tangle parser fuzzer: running %d iterations\n%!" iterations;
  for i = 1 to iterations do
    let input = random_input rng 4096 in
    fuzz_one_input input;
    if i mod 10_000 = 0 then
      Printf.printf "  ... %d iterations complete\n%!" i
  done;
  Printf.printf "Tangle parser fuzzer: %d iterations passed with no crashes\n%!" iterations

File: src/rust/fuzz/fuzz_targets/fuzz_parser.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// SPDX-FileCopyrightText: 2026 Jonathan D.A. Jewell
//
// Fuzz target for the TANGLE (Rust) parser.
//
// Invariant: the parser must NEVER panic on ANY input. It should return
// ParseError, never abort.
//
// TANGLE is a braid/knot theory language with operators for vertical
// composition (.), horizontal tensor (|), pipeline (>>), and braid
// generators. This harness generates structured inputs biased toward
// TANGLE syntax.
//
// Run with:
//   cargo fuzz run fuzz_parser

#![no_main]

use libfuzzer_sys::fuzz_target;

/// TANGLE keywords, operators, and syntax fragments.
const FRAGMENTS: &[&str] = &[
    // Keywords
    "def", "weave", "into", "yield", "strands", "compute", "assert",
    "match", "with", "end", "let", "in", "close", "mirror", "reverse",
    "simplify", "cap", "cup", "twist",
    // Operators
    ".", "|", ">>", "==", "~", "+", "-", "*", "/",
    // Braid generators
    "s1", "s2", "s3", "s1^-1", "s2^-1",
    // Delimiters
    "(", ")", "[", "]", "{", "}", ",", ";", ":",
    "->", "=>",
    // Literals
    "42", "0", "3.14", "true", "false",
    // Identifiers
    "x", "y", "braid", "knot", "link", "_",
    // Whitespace
    " ", "\t", "\n",
    // Braid notation
    "braid[", "]",
    // Comments
    "// comment\n", "/* block */",
];

fn structured_input(data: &[u8]) -> String {
    let mut out = String::with_capacity(data.len() * 4);
    for &b in data {
        let idx = (b as usize) % FRAGMENTS.len();
        out.push_str(FRAGMENTS[idx]);
        if b & 0x80 != 0 {
            out.push(' ');
        }
    }
    out
}

fuzz_target!(|data: &[u8]| {
    // Strategy 1: raw UTF-8
    let raw = String::from_utf8_lossy(data);
    {
        let mut parser = tanglec::parser::Parser::new(&raw);
        let _ = parser.parse_program();
    }

    // Strategy 2: structured token-plausible input
    if data.len() > 2 {
        let structured = structured_input(&data[2..]);
        let mut parser = tanglec::parser::Parser::new(&structured);
        let _ = parser.parse_program();
    }
});
